src/lib/optimizer/strategy/nested_expression_rule.cpp

```diff
@@ -0,0 +1,144 @@
+#include "nested_expression_rule.hpp"
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "constant_mappings.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/lqp_column_reference.hpp"
+#include "logical_query_plan/lqp_expression.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "resolve_type.hpp"
+#include "utils/arithmetic_operator_expression.hpp"
+
+namespace opossum {
+
+std::string NestedExpressionRule::name() const { return "Nested Expression Rule"; }
+
+bool NestedExpressionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) {
+  if (node->type() != LQPNodeType::Projection) {
+    return _apply_to_children(node);
+  }
+
+  auto projection_node = std::dynamic_pointer_cast<ProjectionNode>(node);
+
+  auto column_id = ColumnID{0};
+  for (auto expression : projection_node->column_expressions()) {
+    if (expression->is_arithmetic_operator()) {
+      const auto expression_type = _get_type_of_expression(expression);
+      if (expression_type != std::nullopt) {
+        auto value = NULL_VALUE;
+        resolve_data_type(*expression_type, [&](auto type) { value = _evaluate_expression(type, expression); });
+
+        if (!variant_is_null(value) &&
+            _replace_expression_in_parents(node, node->output_column_references()[column_id], value)) {
+          _remove_column_from_projection(projection_node, column_id);
+        }
+      }
+    }
+
+    column_id++;
+  }
+
+  return _apply_to_children(node);
+}
+
+bool NestedExpressionRule::_replace_expression_in_parents(const std::shared_ptr<AbstractLQPNode>& node,
+                                                          const LQPColumnReference& column_reference,
+                                                          const AllTypeVariant& value) {
+  auto changed = false;
+  for (auto parent : node->parents()) {
+    if (parent->type() == LQPNodeType::Predicate) {
+      auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(parent);
+
+      if (is_lqp_column_reference(predicate_node->value()) &&
```

Comment:
Could you add some more comments to this file? I am not sure what is happening here and why.